### PR TITLE
[DevOps] Snyk Integration

### PR DESCRIPTION
**Pull request by ChathurindaRanasinghe**
In this PR we
- created a new organization in [Snyk](https://app.snyk.io/org/py_data_table). 
- created a new service account for [snyk token](https://app.snyk.io/org/py_data_table/manage/service-accounts).

**PARENT ISSUES**

- https://github.com/h2oai/h2o-ops/issues/241
- https://github.com/h2oai/h2o-ops/issues/248

### CODE REVIEW

**Comment by mmalohlava** [comment]
@ChathurindaRanasinghe thank you! great effort - can you point me to the view in Snyk UI to find it?

**Comment by mmalohlava** [comment]
@ChathurindaRanasinghe @hasithjp i need a naming pattern in Snyk UI to be followed `name of component/release version`

File: .github/workflows/snyk-scan.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+jobs:
+  snyk_scan_test:
+    if: ${{ github.event_name == 'pull_request' }}
+    strategy:
+      matrix:
+        depsfiles: [
+          requirements_docs.txt,
+          requirements_extra.txt,
+        ]
```
Comment (oleksiyskononenko) [inline .github/workflows/snyk-scan.yml:20]: Don't you want to add `requirements_tests.txt` here?

File: .github/workflows/snyk-scan.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        with:
+          args: --command=python3 --package-manager=pip --file=${{ matrix.depsfiles }} --skip-unresolved -d --fail-on=all
+  
+  snyk_scan_monitor:
+    if: ${{ github.event_name == 'push' }}
+    strategy:
+      matrix:
+        depsfiles: [
+          requirements_docs.txt,
+          requirements_extra.txt,
```
Comment (oleksiyskononenko) [inline .github/workflows/snyk-scan.yml:38]: `requirements_tests.txt` here too?

File: .github/workflows/snyk-scan.yml (hunk 1 above)
Comment (ChathurindaRanasinghe) [reply .github/workflows/snyk-scan.yml:20]: Dependencies related to production were considered. Dev dependencies were ignored.

File: .github/workflows/snyk-scan.yml (hunk 2 above)
Comment (ChathurindaRanasinghe) [reply .github/workflows/snyk-scan.yml:38]: Same here, only production dependencies were considered. If we want to scan dev dependencies as well we can add them.

**Comment by ChathurindaRanasinghe** [comment]
> @ChathurindaRanasinghe thank you! great effort - can you point me to the view in Snyk UI to find it?

https://app.snyk.io/org/py_data_table
Screenshot 2023-02-11 at 07 42 19

You can change the organisation to PY-DATA-TABLE in Snyk Dashboard.
Screenshot 2023-02-11 at 07 48 16

**Comment by hasithjp** [comment]
> @ChathurindaRanasinghe @hasithjp i need a naming pattern in Snyk UI to be followed `name of component/release version (or release branch name)`

Yes Michal, we can do something like this since we can go into another granular level grouping since we have the concept of orgs and groups in SNYK now. we'll be creating PRODUCT based Organizations in SNYK now and then we can create project under that PRODUCT Org like COMPONENT/[BRANCH]

File: .github/workflows/snyk-scan.yml (hunk 2 above)
Comment (oleksiyskononenko) [reply .github/workflows/snyk-scan.yml:38]: If it doesn't take a lot of time to scan, it won't hurt to add `requirements_tests.txt`, I guess.

Comment (ChathurindaRanasinghe) [reply .github/workflows/snyk-scan.yml:40]: sure.

File: .github/workflows/snyk-scan.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    steps:
+      - name: Extract github branch/tag name
+        shell: bash
+        run: echo "ref=$(echo ${GITHUB_REF##*/})" >> $GITHUB_OUTPUT
+        id: extract_ref
+      
+      - uses: actions/checkout@master
+
+      - name: Snyk scan for Python dependencies
+        uses: snyk/actions/python-3.7@master
```
[the remark this answers is not in the document]
Comment (ChathurindaRanasinghe) [reply .github/workflows/snyk-scan.yml]: It just scan for Python `3.7`. Changed the version to `3.8`.

File: .github/workflows/snyk-scan.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
+name: Snyk Security Vulnerability Scan
+
+on:
+  pull_request:
+    types: [opened, edited, synchronize]
+  push:
+    tags:
+      - 'v.[0-9]+.[0-100]+.[0-100]+'
```
Comment (hasithjp) [inline .github/workflows/snyk-scan.yml:8]: Please follow this format https://github.com/h2oai/mojo2/blob/974ae7904ae3f2c85a5d16b52ed788f14d361908/.github/workflows/snyk-scan.yml#L4 as mentioned in other PRs

**Review by hasithjp — CHANGES_REQUESTED** [review]

File: .github/workflows/snyk-scan.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply .github/workflows/snyk-scan.yml]: Thanks, makes sense as we will drop support for `3.7` in a couple of months.

File: .github/workflows/snyk-scan.yml (hunk 4 above)
Comment (ChathurindaRanasinghe) [reply .github/workflows/snyk-scan.yml:8]: sure

File: .github/workflows/snyk-scan.yml [hunk 5]
```diff
@@ -0,0 +1,58 @@
+name: Snyk Security Vulnerability Scan
+
+on:
+  pull_request:
+    types: [opened, edited, synchronize]
```
[the remark this answers is not in the document]
Comment (hasithjp) [reply .github/workflows/snyk-scan.yml:6]: please address this

**Review by mmalohlava — APPROVED** [review]
Good to GO, IMHO.

**Comment by ChathurindaRanasinghe** [comment]
> As far as I understand snyk, it looks good to me. Can we test what happens in the case the scan is not passed? I.e. if the build is going to fail and what kind of report snyk will provide.

Example report like [this](https://app.snyk.io/org/py_data_table/reporting?context[page]=issues-detail&project_target=datatable%252Fsnyk-integration&project_origin=cli&issue_status=Open&issue_by=Severity&table_issues_detail_cols=SCORE%257CCVE%257CCWE%257CPROJECT%257CEXPLOIT%2520MATURITY%257CAUTO%2520FIXABLE%257CINTRODUCED%257CSNYK%2520PRODUCT&table_issues_detail_sort=%2520PRIORITY_SCORE%2520DESC)

**Comment by oleksiyskononenko** [comment]
@ChathurindaRanasinghe Thanks, getting "Unable to display this organization".

**Comment by ChathurindaRanasinghe** [comment]
> @ChathurindaRanasinghe Thanks, getting "Unable to display this organization".

Seems like an issue with permission I will look into it.

**Review by oleksiyskononenko — APPROVED** [review]

**Review by hasithjp — APPROVED** [review]

**Pull request merged**